spec/services/teachers/set_funding_eligibilty_spec.rb

```diff
@@ -0,0 +1,110 @@
+        FactoryBot.create(:ect_at_school_period, :ongoing, teacher:)
+      end
+
+      it "sets `first_became_eligible_for_ect_training_at` if not already set" do
+        expect { service.set! }.to change(teacher, :first_became_eligible_for_ect_training_at).from(nil)
+
+        expect(teacher.first_became_eligible_for_ect_training_at).to be_present
+      end
+
+      it "does not change `first_became_eligible_for_ect_training_at` if already set" do
+        teacher.update!(first_became_eligible_for_ect_training_at: 1.day.ago)
+
+        expect { service.set! }.not_to change(teacher, :first_became_eligible_for_ect_training_at)
+      end
+    end
+
+    context "when teacher is not eligible for ECT training" do
+      it "does not set `first_became_eligible_for_ect_training_at`" do
+        expect { service.set! }.not_to change(teacher, :first_became_eligible_for_ect_training_at)
+      end
+    end
+
+    context "when teacher is eligible for mentor training" do
+      it "sets first_became_eligible_for_mentor_training_at if not already set" do
+        expect { service.set! }.to change(teacher, :first_became_eligible_for_mentor_training_at).from(nil)
+
+        expect(teacher.first_became_eligible_for_mentor_training_at).to be_present
+      end
+
+      it "does not `change first_became_eligible_for_mentor_training_at` if already set" do
+        teacher.update!(first_became_eligible_for_mentor_training_at: 1.day.ago)
+
+        expect { service.set! }.not_to change(teacher, :first_became_eligible_for_mentor_training_at)
+      end
+    end
+
+    context "when teacher is not eligible for mentor training" do
+      before do
+        teacher.update!(mentor_became_ineligible_for_funding_on: Time.zone.now, mentor_became_ineligible_for_funding_reason: "completed_declaration_received")
+      end
+
+      it "does not set first_became_eligible_for_mentor_training_at" do
+        expect { service.set! }.not_to change(teacher, :first_became_eligible_for_mentor_training_at)
+      end
+    end
+
+    context "when teacher attributes are changed" do
+      it "records a teacher set funding eligibility event" do
+        expect(Events::Record).to receive(:record_teacher_set_funding_eligibilty_event!)
+          .with(author:, teacher:, happened_at: instance_of(ActiveSupport::TimeWithZone))
+
+        FactoryBot.create(:induction_period, :ongoing, teacher:)
+        FactoryBot.create(:ect_at_school_period, :ongoing, teacher:)
+
+        service.set!
+      end
+    end
+
+    context "when teacher attributes are not changed" do
+      before do
+        teacher.update!(first_became_eligible_for_ect_training_at: Time.zone.now, first_became_eligible_for_mentor_training_at: Time.zone.now)
+      end
+
+      it "does not record a teacher set funding eligibility event" do
+        expect(Events::Record).not_to receive(:record_teacher_set_funding_eligibilty_event!)
+
+        service.set!
+      end
+    end
+
+    it "saves the teacher" do
```

Comment:
I think this and the 'runs inside a transaction' are already covered by your other tests